File: pinochle/card.py
import re
from typing import Optional, Dict

from pinochle import Suit, Rank


class Card:
    """ A card in the pinochle deck """

    def __init__(self, rank: Rank, suit: Suit):
        self.rank: Rank = rank
        self.suit: Suit = suit

    @staticmethod
    def parse(s) -> "Card":
        """ Parses the input string for a card object """
        return CardParser.parse(s)

    def __repr__(self) -> str:
        return self.__str__()

    def __str__(self) -> str:
        """ Gets the english language representation of this card with title case words """
        rankname: str = str(self.rank)
        suitname: str = str(self.suit)
        output: str = f"{rankname} of {suitname}"
        return output

    def __eq__(self, other) -> bool:
        return self.rank == other.rank and self.suit == other.suit

    def __hash__(self) -> int:
        return hash(self.rank) ^ hash(self.suit)


class CardParser:
    """ Parses an input string for a card name """

    @staticmethod
    def parse(s: str) -> Optional[Card]:
        card = None
        m = re.match(r'([A1KQJ9]).* OF ([HCDS]).*', s.upper())
        if m:
            rank_key = m.group(1)
            suit_key = m.group(2)
            rank, suit = rankmap.get(rank_key), suitmap.get(suit_key)
            card = Card(rank, suit)
        return card


rankmap: Dict[str, Rank] = {
    "A": Rank.ACE,
    "1": Rank.TEN,
    "K": Rank.KING,
    "Q": Rank.QUEEN,
    "J": Rank.JACK,
    "9": Rank.NINE
}

suitmap: Dict[str, Suit] = {
    'H': Suit.HEARTS,
    'C': Suit.CLUBS,
    'D': Suit.DIAMONDS,
    'S': Suit.SPADES
}

File: card.go
package pinochle

import (
	"embed"
	"fmt"
	"io"
	"path/filepath"
)

//go:embed svg_playing_cards
var svgImages embed.FS

// ---------------------------------------------------------------------
// Type definitions
// ---------------------------------------------------------------------

// A Card is a combination of a Rank and a Suit
type Card struct {
	Rank Rank
	Suit Suit
}

// ---------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------

// Creates a new Card from a rank and suit
func NewCard(rank Rank, suit Suit) Card {
	p := new(Card)
	p.Rank = rank
	p.Suit = suit
	return *p
}

// ---------------------------------------------------------------------
// Methods
// ---------------------------------------------------------------------

// Returns a representation of the card as a string
func (c Card) String() string {
	s := c.Rank.String() + c.Suit.String()
	return s
}

// Returns a representation of the card as a Unicode string
func (c Card) Unicode() string {
	suitOffset := c.Suit.Offset()
	rankOffset := c.Rank.Offset()
	cardOffset := suitOffset + rankOffset
	s := fmt.Sprintf("%c", cardOffset)
	return s
}

// Returns the SVG image of this card
func (c Card) GetSVG() (string, error) {
	var rankName, suitName string

	// Get the rank name
	switch c.Rank {
	case TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE, TEN:
		rankName = fmt.Sprintf("%d", c.Rank)
	case JACK:
		rankName = "jack"
	case QUEEN:
		rankName = "queen"
	case KING:
		rankName = "king"
	case ACE:
		rankName = "ace"
	}

	// Get the suit name
	switch c.Suit {
	case SPADES:
		suitName = "spades"
	case HEARTS:
		suitName = "hearts"
	case DIAMONDS:
		suitName = "diamonds"
	case CLUBS:
		suitName = "clubs"
	}

	// Create the file name
	baseName := suitName + "_" + rankName + ".svg"
	fileName := filepath.Join("svg_playing_cards", "fronts", baseName)

	// Load the SVG contents
	fp, err := svgImages.Open(fileName)
	if err != nil {
		return "", err
	}
	defer fp.Close()
	contents, _ := io.ReadAll(fp)

	// Return contents as a string
	return string(contents), nil
}
